util/futures/setter.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package futures
+
+import (
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4"
+)
+
+// Setter interface defines methods that an object should implement in order to
+// use the futures package for setting futures.
+type Setter interface {
+	Getter
+	SetFutures(infrav1.Futures)
+}
+
+// Set sets the given future.
+//
+// NOTE: If a future already exists, we update it.
+func Set(to Setter, future *infrav1.Future) {
+	if to == nil || future == nil {
+		return
+	}
+
+	// Check if the new future already exists, and update it if it does.
+	futures := to.GetFutures()
+	exists := false
+	for i, f := range futures {
+		if f.Name == future.Name && f.ServiceName == future.ServiceName {
+			exists = true
+			futures[i] = *future
+		}
+	}
+
+	// If the future does not exist, add it.
+	if !exists {
+		futures = append(futures, *future)
+	}
+
+	to.SetFutures(futures)
+}
+
+// Delete deletes the specified future.
+func Delete(to Setter, name, service string) {
+	if to == nil || name == "" || service == "" {
+		return
+	}
+
+	futures := to.GetFutures()
+	for i := len(futures) - 1; i >= 0; i-- {
+		if futures[i].Name == name && futures[i].ServiceName == service {
+			futures = append(futures[:i], futures[i+1:]...)
```

Comment:
```suggestion
			futures = append(futures[:i], futures[i+1:]...)
			break
```

Reply:
is there a reason for iterating from the end? Is it because we are updating the slice in the loop?

Reply:
It would make sense if we were to remove more than one item from the slice so we don't have to worry about deleted indices but since we're assuming the future is unique and breaking after finding it I don't think it's strictly necessary

Reply:
changed the order